### ISSUE TITLE
Wont run on latest insider build (for WSL 2)

### ISSUE THREAD

**Issue by myoung34** [mintty/wsltty#171]
Worked until latest version of windows (insider build for WSL 2 support)

```
C:\Users\3vilp\AppData\Local\wsltty\bin>ver

Microsoft Windows [Version 10.0.18917.1000]
```

![screen shot here](https://i.imgur.com/v7BSza2.jpg)

**Comment by myoung34** [comment]
Update: reinstalled wsltty but now the error is a bit different:

![](https://i.imgur.com/frfJe07.jpg)

**Comment by donperi** [comment]
Same issue.

**Comment by mintty** [comment]
This way of starting the wsltty-installed copy of mintty was never supposed to run. You would use the shortcut (desktop or start menu) or, from a cmd.exe, the WSL.bat script.

**Comment by myoung34** [comment]
This is from debugging post upgrade.

```
/bin/wslbridge: Exit 1.
wslbridge error: failed to start backend process
note: backend error output: wslbridge-backend: wslbridge-backend.cc:45: int {anonymous}::connectSocket(int, const string&): Assertion `connectRet == 0' failed.

``` 
is what happens when i run it from a .bat

Ive already opened https://github.com/rprichard/wslbridge/issues/44 as a result

**Comment by binkley** [comment]
For me after migrating an Ubuntu distro to WSL 2, wsltty window opens, and closes again.  I'm unsure how to get some error text to examine.

I've tried `wsl.bat` (same behavior), and usual way of running (same behavior) is this desktop shortcut:
```
%LOCALAPPDATA%\wsltty\bin\mintty.exe --WSL= --configdir="%APPDATA%\wsltty" -~ 
```

**Comment by mintty** [comment]
Does "WSL 2" still work for anybody? Would anyone open an issue with WSL about broken interworking?
Is there any MS publication about what they did to break it?

**Comment by mintty** [comment]
Also, does reverting to WSL 1 fix the issue?
`wsl --set-default-version 1`

**Comment by Biswa96** [comment]
WSL2 works in my PC well. @mintty are you using insider builds? There are [many issues](https://github.com/microsoft/WSL) that have to be fixed. I also did break it in first time. Check if DHCP Client, DNS Client and Host Network Service are running properly. This may be worth of reading https://docs.microsoft.com/en-us/windows/wsl/wsl2-ux-changes.

@myoung34 How did you debug wslbridge? Mine only shows `wslbridge error: failed to start backend process`. May you provide a lazy debugging steps (minimal works) for both frontend and backend?

**Comment by myoung34** [comment]
@Biswa96 mine shows that error as well as of yesterday

**Comment by identityope** [comment]
@mintty reverting to WSL 1 will fix the issue

**Comment by mintty** [comment]
I don't have WSL 2 yet (or the required Windows update).
What happens if you run a `cmd` console window and invoke `wsl` there - does it run inside the window as it used to, or does it open a new window (as suggested by the descriptions that they are running in a VM now)?

**Comment by myoung34** [comment]
wsl runs fine from cmd.exe, thats currently the only way to get into wsl 2.
wsltty (namely wslbridge) are just plain broken after the insider update

**Comment by therealkenc** [comment]
The diverge is that WSL2 isn't on localhost as with WSL1, so the [`INADDR_LOOPBACK`](https://github.com/rprichard/wslbridge/blob/47b41bec6c32da58ab01de9345087b1a4fd836e3/frontend/wslbridge.cc#L80) premise wslbridge is based on no longer holds. Worse, the IP address of the vEthernet interface of the VM isn't stable.

**Comment by mintty** [comment]
Is any relevant information in the registry? How would wslbridge connect to its backend anyway?

**Comment by Biswa96** [comment]
We should not depends on registry. It is subjected to change. Ben Hillis warned me twice.

**Comment by therealkenc** [comment]
> Is any relevant information in the registry?

Not to the question at hand. Spitballing (and not thinking about it very hard) one could spawn `wsl.exe ifconfig` and parse the output for the `eth0` address. You can do that on WSL1 too. Maybe as QFE you could use an environment variable (`WSLTTY_IPV4ADDR`) and make the user set it.

WSL2 is (to a first order) the same as a Hyper-V Linux Container on Windows. Imagine hypothetical "mintty for Windows Docker".

**Comment by Biswa96** [comment]
How about using the Hyper-V socket that wsl.exe or bash.exe uses? Although It is not stable and used internally.

**Comment by mintty** [comment]
If you have an idea how to solve the issue, go ahead and try it.
Also I'd appreciate if someone reports a bug to WSL.

**Comment by therealkenc** [comment]
> Also I'd appreciate if someone reports a bug to WSL.

You can track tangential [#3279](https://github.com/microsoft/WSL/issues/3279) although that isn't exactly on point and won't address the problem. To the extent that one flips status it would obviate the need for the bridge.

As of this writing today, the cause of the `wsltty` fail is by-design, analogous to [WSL#4106](https://github.com/microsoft/WSL/issues/4106), because WSL now has its own IP space (indeed by popular request). 

This said (and complicating matters) I gather sharing `127.0.0.1` in WSL2 is being complemented. Landing zone for that will probably end up being [WSL#4199](https://github.com/microsoft/WSL/issues/4199). However, note _if_ that one ever flips status, it will almost certainly be behind a flag. Because sharing localhost breaks everything that _doesn't_ want to share Window's IP space. The killer app being Samba.

**Comment by Biswa96** [comment]
From the hint by @therealkenc

> The diverge is that WSL2 isn't on localhost as with WSL1, so the INADDR_LOOPBACK premise wslbridge is based on no longer holds.

I have tried with HyperV sockets. See my [hvtool project](https://github.com/Biswa96/hvtool). But the problem is that cygwin does not support it, see [this mail](https://cygwin.com/ml/cygwin/2019-07/msg00022.html).

**Comment by Biswa96** [comment]
Anyone interested? Try my [new puppy](https://github.com/Biswa96/wslbridge2/tree/master/rawpty). Open mintty, run `rawpty.exe wsl.exe`. One bug: process does not exit until terminal window is resized. Any help would be appreciated.

**Comment by Biswa96** [comment]
Anyone interested? Try my [new puppy](https://github.com/Biswa96/wslbridge2/tree/master/rawpty). Open mintty, run `rawpty.exe wsl.exe`. That also works with any Win32 executable and WSL1 too. One bug: process does not exit until terminal window is resized. Any help would be appreciated.

**Comment by mintty** [comment]
`conhost --vtmode ...`: Is that the trick to avoid handling the ConPty API directly?

**Comment by Biswa96** [comment]
No. Assume that as alternative of `$TERM`.

**Comment by therealkenc** [comment]
> Also resizing the mintty window will not wrap text (there may be issue about this in mintty).

Not exactly "in mintty" per se. To do this (air quote) "right" you've can use the ConPTY interface and call [`PseudoConsoleResize()`](https://docs.microsoft.com/en-us/windows/console/resizepseudoconsole). I do this [here](https://github.com/therealkenc/wslterm/blob/22be96f397d11dc6f63901911a49405cc3eb8cdd/src/child.c#L1100) in wslterm. 

My motivation (if that's the word) for doing the pty code in a `mintty` fork is that `mintty` was patched upstream  for wslbridge _anyway_and doing it all in `mintty` makes the solution a one-stop-shop. [Full disclosure, I also (incorrectly) thought hacking up `child.c` would be more trivial and a quicker path to victory.] I had a `rawpty` (by another name) in mind too, so you are totally on the right track.

What we're _really_ writing here is a win32 port of [socat](https://linux.die.net/man/1/socat). I call dibs on the github repo names `wslcat` and `winsocat` if I ever get around to writing it. Or Bis, feel _really really_ encouraged to beat me to it.  This is the summer so my time-sink hours are more limited. But anyway, the WSL side of wslbridge in it's present form (I hate using the word 'WSL' like that because it has nothing to do with WSL) is really just `socat` (n.b. this is typed blind off cuff):
```
C:\> winsocat -,raw,echo=0 tcp:localhost:3023 wsl.exe -d Ubuntu-18.04
```
and on the Linux side...
```
$ socat tcp-l:3023,reuseaddr,fork exec:/bin/bin,pty,setsid,setpgid,stderr,ctty &
```

Which is _exactly_, as the wslbridge [readme](https://github.com/rprichard/wslbridge) says, quoth "connecting to the WSL command-line environment over TCP sockets, as with ssh, but without the overhead of configuring an SSH server".

Not that you would do it that way, because we don't need to tunnel (read: telnet) tty over tcp. Again hand-waving:
```
C:\> winsocat -,raw,echo=0 wsl.exe
```
Which is Biswa96's `rawpty`. Or something like that.

And as a closing note, if it isn't obvious, that should ideally be built into `wsl.exe` by the MSFT devs, fixing [WSL#3279](https://github.com/microsoft/WSL/issues/3279), and obviating the need for any this.

**Comment by mintty** [comment]
@therealkenc: I tried your mintty-conpty patch yesterday but it did not work for me.
@Biswa96: I just tried rawpty and it runs WSL. A distribution parameter is missing, and on exit it returns to cygwin only after you resize the terminal. The major problem, however, and probably caused by conpty, is that it intercepts escape sequence handling. Most character attributes do not work, most requests do not work or return conpty's own response, not mintty's, e.g. the primary device attribute: `echo -e '\e[c'`;  `echo -e '\e[>c'`,  `echo -e '\e[18t'` and others do not get a response at all.

**Comment by therealkenc** [comment]
> @therealkenc: I tried your mintty-conpty patch yesterday but it did not work for me.

Alright, appreciate someone trying. I think I have a guess as to the problem. I'll edit this post once it is fixed.

**Comment by mintty** [comment]
@Biswa96, I write:

> The major problem, however, and probably caused by conpty, is that it intercepts escape sequence handling. Most character attributes do not work, ...

Maybe the culprit is the ENABLE_VIRTUAL_TERMINAL_PROCESSING flags, because it enables MS's own handling of escape sequences (https://docs.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences), which is the wrong thing to do when running applications (like WSL) from a dedicated terminal.
See these issues for some background information:
https://github.com/rprichard/winpty/issues/140
https://github.com/rprichard/winpty/issues/89
https://github.com/rprichard/winpty/issues/92

**Comment by Biswa96** [comment]
But those flags can't be enable because std handles are pipes in mintty "terminal". I think I have to test different conhost options combination. Can you write a simple shell script to test those escape code? Also there is a `TerminalState::enterRawMode` function. Can you check if there is any error in `termios` flags?

**Comment by mintty** [comment]
Test and demo file for text attributes attached.
[text-attributes.txt](https://github.com/mintty/wsltty/files/3411080/text-attributes.txt)
There are of course lots of other escapes to be checked, both directions.

I have the impression that if we use conhost, and maybe even if we use ConPTY at all, we are artifically restricted to the set of escapes that MS grants us, by handling them in their own idea of "terminal" implementation. Maybe we should request them to provide a transparent mode, passing all escapes unfiltered.

**Comment by Biswa96** [comment]
MS will not help because my method is [not fully documented][1]. Yes, `cat`-ing your provided text shows difference in ConPTY. I am not well versed in escape codes. If you have mingw-w64-gcc tootchain or visual studio build tools you can test my [XConPty][2] project with `make CCOPT=-DFUN_MODE` command. It will [show the underlying window][3] with those escape code, i.e. that removes headless mode from ConPTY.

Another thing, rawpty can now pass all the `wsl.exe` options.

[1]: https://github.com/microsoft/terminal/issues/2010
[2]: https://github.com/Biswa96/XConPty
[3]: https://github.com/Biswa96/XConPty/blob/master/images/Ping_Fun_Mode.PNG

**Comment by Biswa96** [comment]
Though this is not the only way to handle WSL2. I have think of making something with Hyper-V socket + `/dev/pty/n` + telnet/ssh like mixture.

**Comment by mintty** [comment]
> Maybe we should request them to provide a transparent mode, passing all escapes unfiltered.

Actually, that is https://github.com/microsoft/terminal/issues/1173

**Comment by mintty** [comment]
I guess, access via ssh would work well, but you'd need to setup an ssh server first, in every WSL distribution, right? And how do you find the IP/port to connect to?

**Comment by ghs2015** [comment]
I can run the wsltty for WSL2 out of box.

**Comment by therealkenc** [comment]
> the primary device attribute: echo -e '\e[c'; echo -e '\e[>c', echo -e '\e[18t' and others do not get a response at all.

Inspired observation, so thanks for that mintty. That behavior can be framed as bug, Microsoft code all the way down, no Cygwin/MSY2 in sight. I've opened [microsoft/WSL#4327](https://github.com/microsoft/WSL/issues/4327). We'll have to see if it is addressed or goes chirping crickets (I have no special powers on that).

**Comment by Biswa96** [comment]
WSL devs had change the networking to localhost. See this [blog post][1].

[1]: https://devblogs.microsoft.com/commandline/whats-new-for-wsl-in-insiders-preview-build-18945/

**Comment by AiOO** [comment]
I have same issue on Windows 10 18945.1001 & WSLtty 3.0.2. Unfortunately, reverting to WSL 1 can't fix issue. 😢

**Comment by dafa321** [comment]
wsl --set-version <Distro> 1 was failed first time for me when from production to insider program,
 since  then  has been working switch between wsl 1 & 2 or latest 18945.1001 inside update .
just make sure do a back up

**Comment by Biswa96** [comment]
I have added another way to connect to WSL2 from mintty/cygwin. Code here [hvpty](https://github.com/Biswa96/wslbridge2/tree/master/hvpty).

**Comment by Biswa96** [comment]
Now [wslbridge2/hvpty](https://github.com/Biswa96/wslbridge2/tree/master/hvpty) works with mintty and WSL2. Though there are some caveats I will fix it ASAP!

**Comment by Biswa96** [comment]
Cygwin now has Windows ConPTY support. See [this commit][1]. Released in Cygwin 3.1.0-0.3 test build. So, `wsl.exe` and other Windows console programs can be executed directly in mintty, no need to use wslbridge or winpty related programs. 

[1]: https://sourceware.org/git/?p=newlib-cygwin.git;a=commit;h=169d65a5774acc76ce3f3feeedcbae7405aa9b57

**Comment by mintty** [comment]
> wsl.exe and other Windows console programs can be executed directly in mintty, no need to use wslbridge or winpty related programs.

I had tried that already and yes, it works. Unfortunately, this way `wsl` enforces the restricted MS terminal model, i.e. it interprets escape sequences on itw own and the result is not really mintty anymore.
I expect this problem will be resolved once MS gives us conpty "pass-through" mode.

**Comment by Biswa96** [comment]
But this is only test build. Somehow it doesn't work in Windows insider build 18970 but works in 18945.

**Comment by mintty** [comment]
@Biswa96, does either wslbridge or wslbridge2 work for you with WSL 2 as one of your comments above suggests?
According to https://github.com/microsoft/terminal/issues/1173#issuecomment-527275814, WSL 2 does not change the way it runs in the Windows console. However, that statement is based on the assumption that grabbing the console would use `ReadConsoleOutput` which wslbridge does not do in fact, checking the sources. So I wonder how wslbridge actually works; how does it retrieve application output and why would *that* not work with WSL 2 anymore?

**Comment by mintty** [comment]
@Biswa96, in your new tool suite, `hvpty` apparently works as a gateway to WSL 2.
How does it achieve that? What's the difference to wslbridge2, what's the difference to rawpty?
Could wslbridge2 and hvpty be united so they work in common for both WSL 1 and WSL 2?

**Comment by Biswa96** [comment]
> Could wslbridge2 and hvpty be united so they work in common for both WSL 1 and WSL 2?

Just provide a code sample about how to distinguish WSL1 and WSL2 **without reading registry and any undocumented methods**. There are many ways to do that with undocumented methods.

The rest of the questions are too broad. I will try to add the working procedure in project's README.

**Comment by mintty** [comment]
What undocumented methods are you aware of? I wouldn't mind reading the registry, the `Flags` key seems to have the indication (bit 3, mask 0x80).
As a fallback, scanning the output of `wsl -l -v` would work.

**Comment by mintty** [comment]
Mintty already scans the registry for option `--WSL` anyway. I would otherwise integrate the distinction there.

**Comment by mintty** [comment]
Unfortunately, as I said above (https://github.com/mintty/wsltty/issues/171#issuecomment-512584708), rawpty does not connect WSL to mintty functionality transparently; rather it restricts the environment to the MS terminal idea. But wslbridge2 and hvpty do not do that. What's the difference?

**Comment by Biswa96** [comment]
Oooo... Now I understand your query. Remember, once I said about the idea of making a telnet + ssh like thing. _wslbridge2_ uses AF_INET and _hvpty_ uses AF_HYPERV. Both uses their own backend to spawn a master/slave pty in WSL then sucks the raw buffer from slave side through sockets. See this [beautiful piece of art](https://unix.stackexchange.com/a/158604/336403). Whereas _rawpty_ uses the ConPTY implementation in a "raw" way.

Also I have tested the [text-attributes.txt](https://github.com/mintty/wsltty/files/3411080/text-attributes.txt) file that you had provided before.

**Comment by mintty** [comment]
Can you add the `-l` option also to hvpty, please?
Also, not to break #156 again, static linking of the backend will be good.
It will require my getpwuid patch, though.

**Comment by mintty** [comment]
Furthermore, wslbridge2 does not support the special `-C~` parameter value.
Also, hvpty does not have options -t (maybe not needed) and -e. -C~ not checked.

**Comment by Biswa96** [comment]
Can you make a issue in my repo? This issue is getting too long... Finger hurts...

**Comment by mintty** [comment]
Released 3.0.5.

**Comment by mjbright** [comment]
I frankly didn't understand a lot of what was discussed here, but came here as I seemed to have the same problem on WSL2.

I'm on OS build [Version 10.0.19041.113].
This is the first time I've used ttyd in quite a while, I downloaded 1.6.0 and was seeing "Connection Closed".

So I just ran ttyd as
```ttyd -p 1111 wsl.exe bash```
and this worked fine.

I maybe missed the point of this discussion, but in case this simple workaround helps someone ...
